dm/syncer/opt_sharding_group.go

```diff
@@ -0,0 +1,194 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package syncer
+
+import (
+	"sync"
+
+	"github.com/pingcap/tidb/util/filter"
+	"go.uber.org/zap"
+
+	"github.com/pingcap/tiflow/dm/dm/config"
+	"github.com/pingcap/tiflow/dm/pkg/binlog"
+	tcontext "github.com/pingcap/tiflow/dm/pkg/context"
+	"github.com/pingcap/tiflow/dm/pkg/utils"
+)
+
+// OptShardingGroup represents a optimistic sharding DDL sync group.
+type OptShardingGroup struct {
+	sync.RWMutex
+
+	// the conflict tableIDs hash set to quickly check whether this table is in conflict stage
+	// sourceTableID -> each table's conflicted ddl's lastLocation
+	conflictTables map[string]binlog.Location
+
+	firstConflictLocation binlog.Location
+	flavor                string
+	enableGTID            bool
+}
+
+func NewOptShardingGroup(firstConflictLocation binlog.Location, flavor string, enableGTID bool) *OptShardingGroup {
+	return &OptShardingGroup{
+		firstConflictLocation: firstConflictLocation,
+		flavor:                flavor,
+		enableGTID:            enableGTID,
+		conflictTables:        make(map[string]binlog.Location, 1),
```

Comment:
Why `conflictTables` has to have at least 1 cap but set empty while being reset?

Reply:
reset function is never used and deleted now. We always put a table in `conflictTables` so length is set to 1 here.